packages/server/modules/workspaces/tests/unit/services/management.spec.ts

```diff
@@ -0,0 +1,353 @@
+      const eventData = {
+        isCalled: false,
+        eventName: '',
+        payload: {}
+      }
+
+      const userId = cryptoRandomString({ length: 10 })
+      const workspaceId = cryptoRandomString({ length: 10 })
+
+      const role: WorkspaceAcl = { userId, workspaceId, role: Roles.Workspace.Member }
+
+      const storedRoles: WorkspaceAcl[] = [role]
+
+      const deleteWorkspaceRole = deleteWorkspaceRoleFactory({
+        getWorkspaceRoles: async () => storedRoles,
+        deleteWorkspaceRole: async () => {
+          return storedRoles[0]
+        },
+        emitWorkspaceEvent: async ({ eventName, payload }) => {
+          eventData.isCalled = true
+          eventData.eventName = eventName
+          eventData.payload = payload
+
+          return []
+        },
+        getStreams: async () => ({ streams: [], totalCount: 0, cursorDate: null }),
+        revokeStreamPermissions: async () => ({} as StreamRecord)
+      })
+
+      await deleteWorkspaceRole({ userId, workspaceId })
+
+      expect(eventData.isCalled).to.be.true
+      expect(eventData.eventName).to.equal(WorkspaceEvents.RoleDeleted)
+      expect(eventData.payload).to.deep.equal(role)
+    })
+    it('throws if attempting to delete the last admin from a workspace', async () => {
+      const userId = cryptoRandomString({ length: 10 })
+      const workspaceId = cryptoRandomString({ length: 10 })
+
+      const role: WorkspaceAcl = { userId, workspaceId, role: Roles.Workspace.Admin }
+
+      let storedRoles: WorkspaceAcl[] = [role]
+
+      const deleteWorkspaceRole = deleteWorkspaceRoleFactory({
+        getWorkspaceRoles: async () => storedRoles,
+        deleteWorkspaceRole: async ({ userId, workspaceId }) => {
+          const role = storedRoles.find(
+            (r) => r.userId === userId && r.workspaceId === workspaceId
+          )
+
+          storedRoles = storedRoles.filter((r) => r.userId !== userId)
+
+          return role ?? null
+        },
+        emitWorkspaceEvent: async () => [],
+        getStreams: async () => ({ streams: [], totalCount: 0, cursorDate: null }),
+        revokeStreamPermissions: async () => ({} as StreamRecord)
+      })
+
+      await expectToThrow(() => deleteWorkspaceRole({ userId, workspaceId }))
+    })
+    it('deletes workspace project roles', async () => {
+      const userId = cryptoRandomString({ length: 10 })
+      const workspaceId = cryptoRandomString({ length: 10 })
+      const projectId = cryptoRandomString({ length: 10 })
+
+      const workspaceRole: WorkspaceAcl = {
+        userId,
+        workspaceId,
+        role: Roles.Workspace.Member
+      }
+      const workspaceRoles: WorkspaceAcl[] = [workspaceRole]
+      const workspaceProjects: StreamRecord[] = [{ id: projectId } as StreamRecord]
+
+      let projectRoles: StreamAclRecord[] = [
+        { userId, role: Roles.Stream.Contributor, resourceId: projectId }
+      ]
+
+      const deleteWorkspaceRole = deleteWorkspaceRoleFactory({
```

Comment:
Figuring out how DRY tests should be is an art form in and of itself - if you go too DRY you can run into way too abstract tests that are too hard to understand and change (e.g. `modules/comments/tests/comments.graph.spec.js`). That said, copy and pasting the same setup over and over again when it's expected to be the same for most if not all test cases, even if it does change in the future, is going too far in the other direction - just changing the deps a little bit for the role factory will require the dev to go and make a dozen identical changes to all factory invocations across this test file.

I definitely want to push back about this assumption that DRY is not a concern in tests - it most definitely is, you just should not go overboard with it and make it "too clever" and you have to evaluate which bits of code are duplicate only currently and very likely to differ in the future and which are more or less always going to be the same. In this case there's so much duplication of common init, it could easily be extracted to a helper at the top of the file. In the case of factory init, you can also always provide a param that allows you to override some of the dependencies, if needed, while all of the other ones stay standard.

For examples on how I've approached this you can check `automations.spec.ts` or `trigger.spec.ts`

Reply:
(discussed in person but some late thoughts) I think this has landed somewhere much better, lmk what you think, and ty for the push